pkg/config/http_test.go

```diff
@@ -161,3 +164,237 @@ func TestHttpLoader_GetRuntimeConfig(t *testing.T) {
 		})
 	}
 }
+
+// TestHttpLoader_Run tests the Run method of the HttpLoader
+// The test runs the Run method for a while
+// and then shuts it down via a goroutine
+func TestHttpLoader_Run(t *testing.T) {
+	tests := []struct {
+		name     string
+		response *RuntimeConfig
+		code     int
+		wantErr  bool
+	}{
+		{
+			name:     "non-200 response",
+			response: nil,
+			code:     500,
+			wantErr:  false,
+		},
+		{
+			name: "nil checks",
+			response: &RuntimeConfig{
+				Checks: nil,
+			},
+			code: 200,
+		},
+		{
+			name: "empty config",
+			response: &RuntimeConfig{
+				Checks: map[string]any{},
+			},
+			code: 200,
+		},
+		{
+			name: "config with checks",
+			response: &RuntimeConfig{
+				Checks: map[string]any{
+					"testCheck1": map[string]any{
+						"enabled": true,
+					},
+				},
+			},
+			code: 200,
+		},
+	}
+
+	httpmock.Activate()
+	t.Cleanup(httpmock.DeactivateAndReset)
+
+	for _, tt := range tests {
+		resp, err := httpmock.NewJsonResponder(tt.code, tt.response)
+		if err != nil {
+			t.Fatalf("Failed creating json responder: %v", err)
+		}
+
+		httpmock.RegisterResponder("GET", "https://api.test.com/test", resp)
+
+		t.Run(tt.name, func(t *testing.T) {
+			hl := &HttpLoader{
+				cfg: &Config{
+					Loader: LoaderConfig{
+						Type:     "http",
+						Interval: time.Millisecond * 500,
+						Http: HttpLoaderConfig{
+							Url: "https://api.test.com/test",
+							RetryCfg: helper.RetryConfig{
+								Count: 3,
+								Delay: 100 * time.Millisecond,
+							},
+						},
+					},
+				},
+				cCfgChecks: make(chan<- map[string]any, 1),
+				client:     http.DefaultClient,
+				done:       make(chan struct{}, 1),
+			}
+
+			// shutdown routine
+			ctx := context.Background()
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				time.Sleep(time.Millisecond * 600)
+				t.Log("Shutting down the Run method")
+				hl.Shutdown(ctx)
+			}()
+
+			err := hl.Run(ctx)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("HttpLoader.Run() error = %v, wantErr %v", err, tt.wantErr)
+			}
```

Comment:
I'll try. I battled a lot with the tests till I got them running without any race conditions. 1 min for the change, 1h 30m for the ~~tears~~ tests.

Reply:
I've added the httpmock.Reset(); I don't think it's necessary. The whole race-condition-fiasco happens because the blocking routines (like the checks checking a remote endpoint or the loader trying to grab the config from a remote endpoint) are still getting executed, when the tests end and their `Run`  routine is still running (it was never properly terminated). httpmock tries to deactivate the mocks by accessing the transport, but it's still under use from the `Retry` function -> race condition.